packages/peregrine/lib/talons/CategoryTree/useCategoryTree.js

```diff
@@ -0,0 +1,62 @@
+import { useEffect, useMemo } from 'react';
+import { useQuery } from '../../hooks/useQuery';
+
+/**
+ * @typedef {object} CategoryNode
+ * @prop {object} category - category data
+ * @prop {boolean} isLeaf - true if the category has no children
+ */
+
+/**
+ * @typedef { import("graphql").DocumentNode } DocumentNode
+ */
+
+/**
+ * Returns props necessary to render a CategoryTree component.
+ *
+ * @param {object} props
+ * @param {object} props.categories - all fetched categories
+ * @param {number} props.categoryId - category id for this node
+ * @param {DocumentNode} props.query - GraphQL query
+ * @param {function} props.updateCategories - bound action creator
+ * @return {{ childCategories: Map<number, CategoryNode> }}
+ */
+export const useCategoryTree = props => {
+    const { categories, categoryId, query, updateCategories } = props;
+
+    const [queryResult, queryApi] = useQuery(query);
+    const { data } = queryResult;
+    const { runQuery } = queryApi;
+
+    // fetch categories
+    useEffect(() => {
+        if (categoryId != null) {
+            runQuery({ variables: { id: categoryId } });
+        }
+    }, [categoryId, runQuery]);
+
+    // update redux with fetched categories
+    useEffect(() => {
+        if (data && data.category) {
+            updateCategories(data.category);
+        }
+    }, [data, updateCategories]);
+
+    const rootCategory = categories[categoryId];
+    const { children } = rootCategory || {};
+
+    const childCategories = useMemo(() => {
+        const childCategories = new Map();
+
+        for (const id of children || '') {
+            const category = categories[id];
+            const isLeaf = category.children_count === '0';
+
+            childCategories.set(id, { category, isLeaf });
+        }
+
+        return childCategories;
+    }, [categories, children]);
+
+    return { childCategories };
```

Comment:
I'm a big fan of not returning the pure `data` object of these query calls as I think I've done in some of my other talon PRs. This is good and I'll be using this going forward.